test/opt/test_unify_const.cpp

```diff
@@ -0,0 +1,880 @@
+            "OpName %outer \"outer_duplicate\"",
+          },
+        },
+        // composite type null constants. Null constants and zero-valued
+        // constants should not be used to replace each other.
+        {
+          // preserved constants
+          {
+            "%bool_zero = OpConstantFalse %bool",
+            "%float_zero = OpConstant %float 0",
+            "%int_null = OpConstantNull %int",
+            "%double_null = OpConstantNull %double",
+            // inner_struct type null constant.
+            "%null_inner = OpConstantNull %inner_struct",
+            // zero-valued composite constant built from zero-valued constant
+            // component. inner_zero should not be replace by null_inner.
+            "%inner_zero = OpConstantComposite %inner_struct %bool_zero %float_zero",
+            // zero-valued composite contant built from zero-valued constants
+            // and null constants.
+            "%outer_zero = OpConstantComposite %outer_struct %inner_zero %int_null %double_null",
+            // outer_struct type null constant, it should not be replaced by
+            // outer_zero.
+            "%null_outer = OpConstantNull %outer_struct",
+          },
+          // use preserved constants in main
+          {
+            "%inner_var = OpVariable %_pf_inner_struct Function",
+            "OpStore %inner_var %inner_zero",
+            "OpStore %inner_var %null_inner",
+            "%outer_var = OpVariable %_pf_outer_struct Function",
+            "OpStore %outer_var %outer_zero",
+            "OpStore %outer_var %null_outer",
+          },
+          // duplicated constants
+          {
+            "%null_inner_dup = OpConstantNull %inner_struct",
+            "%null_outer_dup = OpConstantNull %outer_struct",
+            "%inner_zero_dup = OpConstantComposite %inner_struct %bool_zero %float_zero",
+            "%outer_zero_dup = OpConstantComposite %outer_struct %inner_zero_dup %int_null %double_null",
+          },
+          // use duplicated constants in main
+          {
+            "%inner_var = OpVariable %_pf_inner_struct Function",
+            "OpStore %inner_var %inner_zero_dup",
+            "OpStore %inner_var %null_inner_dup",
+            "%outer_var = OpVariable %_pf_outer_struct Function",
+            "OpStore %outer_var %outer_zero_dup",
+            "OpStore %outer_var %null_outer_dup",
+          },
+          // remapped names
+          {
+            "OpName %null_inner \"null_inner_dup\"",
+            "OpName %null_outer \"null_outer_dup\"",
+            "OpName %inner_zero \"inner_zero_dup\"",
+            "OpName %outer_zero \"outer_zero_dup\"",
+          },
+        },
+        // Spec Constants with SpecId decortion should be skipped.
+        {
+          // preserved constants
+          {
+            // Assembly builder will add OpDecorate SpecId instruction for the
+            // following spec constant instructions automatically.
```

Comment:
That automatic part means this test is a bit fragile....  If the assembly builder changes then this test may silently break.
 Avoid creeping complexity like this.